### PR TITLE
Global refactor

### PR DESCRIPTION
**Pull request by kamil506**
@gotar

### CODE REVIEW

File: lib/db.rb [hunk 1]
```diff
@@ -5,10 +5,10 @@ class Database
 
   def initialize
     @p1 = Product.new("Ball", 13, 8)
-    @p2 = Product.new("asd", 123.15, 8)
+    @p2 = Product.new("Bread", 2.15, 8)
     @p3 = Product.new("Beef", 19.15, 23)
     @p4 = Product.new("Milk", 54.15, 23)
-    @p5 = Product.new("Cheese", 12.54, 9)
+    @p5 = Product.new("Cheese", 12.54, 7)
```
Comment (gotar) [inline lib/db.rb]: will be better to use more descriptive names, like `product`, not `p`. Plus looks like you don't need instance veriables, local will be enough (without `@`)

File: lib/product_holder.rb [hunk 2]
```diff
@@ -1,33 +1,47 @@
...
-  	@products = []
+class ProductHolder
+  attr_reader :products
+
+  def initialize
+    @products = []
+  end
+
+  def fetch_product(product)
+    @found_product = @products.find { |product| product[:product] == product }
```
Comment (gotar) [inline lib/product_holder.rb]: you have reader on products, so you don't need `@`

File: lib/product_holder.rb [hunk 3]
```diff
@@ -1,33 +1,47 @@
...
-      end
+
+    @found_product = fetch_product(product)
+
+    if @found_product
+      @found_product[:value] += 1
+      return @products
     end
+
     @products << {product: product, value: value}
```
Comment (gotar) [inline lib/product_holder.rb]: more clean (ruby way) will be to use `else` block here

File: lib/basket.rb [hunk 4]
```diff
@@ -1,34 +1,31 @@
...
   def bill_netto
-    @products.reduce(0) {|memo, x| 
-      memo += x[:product].price * x[:value]}
+    @products.reduce(0) {|memo, x| memo += x[:product].price * x[:value]}
   end
 
   def bill_brutto
-    @products.reduce(0) {|memo, x| 
-      memo += (x[:product].price + 
+    @products.reduce(0) {|memo, x|
```
Comment (mmarcinkowski) [inline lib/basket.rb]: please use multiline syntax instead 
```
... do |block_var|
  ...
end
```
also please replace `x` with more relevant name

File: lib/product_holder.rb [hunk 5]
```diff
@@ -1,33 +1,47 @@
...
+        stop
+      elsif @found_product[:value] == value
+        @products.delete(@found_product)
+      else
+        @found_product[:value] -= value
       end
+      return @products
     end
+
     raise ArgumentError
```
Comment (gotar) [inline lib/product_holder.rb]: here too, can be written in more clean way. those 2 level if blocks and raise at end are not easy to read

File: lib/warehouse.rb [hunk 6]
```diff
@@ -1,34 +1,19 @@
...
-    @products.each do |x|
-      if x[:product].id == id
-        x[:value] += 1
-      end
-    end
-  end
-
   def avaible?(product, value)
-    @products.each do |x| 
+    @products.each do |x|
```
Comment (gotar) [inline lib/warehouse.rb]: you can replace it by `any?` method

File: spec/unit/product_holder_spec.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
+require_relative "../../lib/product_holder"
+
+RSpec.describe ProductHolder do
+  subject(:pholder) { ProductHolder.new }
+  let(:product1) { Product.new("foo", 15 , 5) }
+  let(:product2) { Product.new("bar", 50 , 8) }
+
+  context("#add") do
+    it "adding proper hash to array" do
+      expect (pholder.add(product1, 10))[0].eql?({:product => product1, :value => 10})
```
Comment (gotar) [inline spec/unit/product_holder_spec.rb]: you do not test anything here, missing `.to` part

File: spec/unit/shop_spec.rb [hunk 8]
```diff
@@ -23,23 +23,20 @@
...
-    expect(shop.find_by_id(1)).to eql(@p)
-  end
-
-  it "return nil for not existing product" do
-    expect(shop.find_by_id(7)).to eql(nil)
-  end
+    it "proper return existing product" do
+      expect(shop.find(1)).to eql(@p)
+    end
+    it "return nil for not existing product" do
```
Comment (gotar) [inline spec/unit/shop_spec.rb]: add empty line between tests

File: lib/basket.rb [hunk 9]
```diff
@@ -1,34 +1,31 @@
...
   attr_reader :products
-  
-  def initialize
-    init
-  end
 
   def bill_netto
-    @products.reduce(0) {|memo, x| 
-      memo += x[:product].price * x[:value]}
+    @products.reduce(0) {|memo, x| memo += x[:product].price * x[:value]}
```
Comment (mmarcinkowski) [inline lib/basket.rb]: please add additional spaces inside the block

**Comment by gotar** [comment]
"Fixing code, added sample views in Sinatra" this commit should be split, probably to 2 separate one, or even more. One commit should contain changes only related to one task, like:
- refactor - one commit
- added Sinatra ... - second one
- ...

**Pull request merged**